File: matssocket-client-javascript/test/integration_push.js
// Register as an UMD module - source: https://github.com/umdjs/umd/blob/master/templates/commonjsStrict.js
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define(['chai', 'sinon', 'ws', 'mats', 'env'], factory);
    } else if (typeof exports === 'object' && typeof exports.nodeName !== 'string') {
        // CommonJS
        const chai = require('chai');
        const sinon = require('sinon');
        const ws = require('ws');
        const mats = require('../lib/MatsSocket');

        factory(chai, sinon, ws, mats, process.env);
    } else {
        // Browser globals
        factory(chai, sinon, WebSocket, mats, {});
    }
}(typeof self !== 'undefined' ? self : this, function (chai, sinon, ws, mats, env) {
    const MatsSocket = mats.MatsSocket;

    describe('MatsSocket integration tests of Server-side send/request ("push")', function () {
        let matsSocket;

        function setAuth(userId = "standard", duration = 20000, roomForLatencyMillis = 10000) {
            const now = Date.now();
            const expiry = now + duration;
            matsSocket.setCurrentAuthorization("DummyAuth:" + userId + ":" + expiry, expiry, roomForLatencyMillis);
        }

        const urls = env.MATS_SOCKET_URLS || "ws://localhost:8080/matssocket,ws://localhost:8081/matssocket";

        beforeEach(() => {
            matsSocket = new MatsSocket("TestApp", "1.2.3", urls.split(","));
            matsSocket.logging = false;
        });

        afterEach(() => {
            // :: Chill the close slightly, so as to get the final "ACK2" envelope over to delete server's inbox.
            // NOTE: This is done async, so for the fast tests, the closings will come in "behind".
            let toClose = matsSocket;
            setTimeout(function () {
                toClose.close("Test done");
            }, 500);
        });

        describe('MatsSocketServer.send()', function () {
            // Set a valid authorization before each request
            beforeEach(() => setAuth());

            it('Send a message to Server, which responds by sending a message to terminator at Client (us!), directly in the MatsStage', function (done) {
                let traceId = "MatsSocketServer.send_test_" + matsSocket.id(6);

                matsSocket.terminator("ClientSide.terminator", (msg) => {
                    chai.assert.strictEqual(msg.data.number, Math.E);
                    chai.assert.strictEqual(msg.traceId, traceId + ":SentFromMatsStage");
                    done()
                });
                matsSocket.send("Test.server.send.matsStage", traceId, {
                    number: Math.E
                })
            });

            it('Send a message to Server, which responds by sending a message to terminator at Client (us!), in a separate Thread', function (done) {
                let traceId = "MatsSocketServer.send_test_" + matsSocket.id(6);

                matsSocket.terminator("ClientSide.terminator", (msg) => {
                    chai.assert.strictEqual(msg.data.number, Math.E);
                    chai.assert.strictEqual(msg.traceId, traceId + ":SentFromThread");
                    done()
                });
                matsSocket.send("Test.server.send.thread", traceId, {
                    number: Math.E
                })
            });
        });

        describe('MatsSocketServer.request()', function () {
            // Set a valid authorization before each request
            beforeEach(() => setAuth());

            function doTest(startEndpoint, resolveReject, done) {
                let traceId = "MatsSocketServer.send_test_" + matsSocket.id(6);

                let initialMessage = "Message_" + matsSocket.id(20);

                // This endpoint will get a request from the Server, to which we respond - and the server will then send the reply back to the Terminator below.
                matsSocket.endpoint("ClientSide.endpoint", function (messageEvent) {
                    chai.assert.strictEqual(messageEvent.traceId, traceId);

                    return new Promise(function (resolve, reject) {
                        // Resolve it a tad later, to "emulate" some kind of processing
                        setTimeout(function () {
                            let data = messageEvent.data;
                            let msg = {
                                string: data.string + ":From_IntegrationEndpointA",
                                number: data.number + Math.PI
                            };
                            // ?: Resolve or Reject based on param
                            if (resolveReject) {
                                resolve(msg);
                            } else {
                                reject(msg);
                            }

                        }, 25);
                    });
                });

                // This terminator will get the final result from the Server
                matsSocket.terminator("ClientSide.terminator", (msg) => {
                    chai.assert.strictEqual(msg.data.number, Math.E + Math.PI);
                    chai.assert.strictEqual(msg.data.string, initialMessage + ":From_IntegrationEndpointA:" + (resolveReject ? "RESOLVE" : "REJECT"));
                    chai.assert.strictEqual(msg.traceId, traceId);
                    done()
                });

                // Here we send the message that starts the cascade
                matsSocket.send(startEndpoint, traceId, {
                    string: initialMessage,
                    number: Math.E
                });
            }

            it('Send a message to the Server, which responds by directly doing a Server-to-Client request (thus coming back here!), resolves, and when this returns to Server, sends it back to a Client Terminator', function (done) {
                doTest("Test.server.request.direct", true, done);
            });

            it('Send a message to the Server, which responds by directly doing a Server-to-Client request (thus coming back here!), rejects, and when this returns to Server, sends it back to a Client Terminator', function (done) {
                doTest("Test.server.request.direct", false, done);
            });

            it('Send a message to the Server, which responds by - in a Mats terminator - doing a Server-to-Client request (thus coming back here!), resolve, and when this returns to Server, sends it back to a Client Terminator', function (done) {
                doTest("Test.server.request.viaMats", true, done);
            });

            it('Send a message to the Server, which responds by - in a Mats terminator - doing a Server-to-Client request (thus coming back here!), rejects, and when this returns to Server, sends it back to a Client Terminator', function (done) {
                doTest("Test.server.request.viaMats", false, done);
            });
        });

        describe('Server initiated sends and requests - with DebugOptions', function () {
            function extracted(msg, debugOptions) {
                // This is a server-initiated message, so these should be undefined in MessageEvent
                chai.assert.isUndefined(msg.clientRequestTimestamp);
                chai.assert.isUndefined(msg.roundTripMillis);

                // :: Assert debug
                // It should be present in MessageEvent, since it should be in the message from the server
                chai.assert.isDefined(msg.debug);
                // These should be set
                if ((debugOptions & mats.DebugOption.TIMESTAMPS) > 0) {
                    chai.assert.isNumber(msg.debug.serverMessageCreated);
                    chai.assert.isNumber(msg.debug.messageSentToClient);
                } else {
                    chai.assert.isUndefined(msg.debug.serverMessageCreated);
                    chai.assert.isUndefined(msg.debug.messageSentToClient);
                }
                if ((debugOptions & mats.DebugOption.NODES) > 0) {
                    chai.assert.isDefined(msg.debug.serverMessageCreatedNodename);
                    chai.assert.isDefined(msg.debug.messageSentToClientNodename);
                } else {
                    chai.assert.isUndefined(msg.debug.serverMessageCreatedNodename);
                    chai.assert.isUndefined(msg.debug.messageSentToClientNodename);
                }
                // While all other should not be set
                chai.assert.isUndefined(msg.debug.clientMessageSent);
                chai.assert.isUndefined(msg.debug.clientMessageReceived);
                chai.assert.isUndefined(msg.debug.clientMessageReceivedNodename);
                chai.assert.isUndefined(msg.debug.matsMessageSent);
                chai.assert.isUndefined(msg.debug.matsMessageReplyReceived);
                chai.assert.isUndefined(msg.debug.matsMessageReplyReceivedNodename);

                // This is not a request from Client, so there is no requested debug options.
                chai.assert.isUndefined(msg.debug.requestedDebugOptions);
                // This is the resolved from what we asked server to use above, and what we are allowed to.
                chai.assert.equal(msg.debug.resolvedDebugOptions, matsSocket.debug);
            }

            function testServerInitiatedSend_DebugOptions(debugOptions, done) {
                // Set special userId that gives us all DebugOptions
                setAuth('enableAllDebugOptions');

                let traceId = "MatsSocketServer.DebugOptions_server.send_" + matsSocket.id(6);

                // These will become the server's initiation requested DebugOptions upon the subsequent 'send'
                matsSocket.debug = debugOptions;

                matsSocket.terminator("ClientSide.terminator", (msg) => {
                    chai.assert.strictEqual(msg.data.number, Math.E);
                    chai.assert.strictEqual(msg.traceId, traceId + ":SentFromThread");

                    extracted(msg, debugOptions);
                    done()
                });

                matsSocket.send("Test.server.send.thread", traceId, {
                    number: Math.E
                })
            }

            it('Server initiated send should have debug object if user asks for it - with the all server-sent stuff filled.', function (done) {
                testServerInitiatedSend_DebugOptions(mats.DebugOption.NODES | mats.DebugOption.TIMESTAMPS, done);
            });
            it('Server initiated send should have debug object if user asks for just nodes - with just the nodes filled.', function (done) {
                testServerInitiatedSend_DebugOptions(mats.DebugOption.NODES, done);
            });
            it('Server initiated send should have debug object if user asks for just timestamps - with just the timestamps filled.', function (done) {
                testServerInitiatedSend_DebugOptions(mats.DebugOption.TIMESTAMPS, done);
            });

            function testServerInitiatedRequest_DebugOptions(debugOptions, done) {
                // Set special userId that gives us all DebugOptions
                setAuth('enableAllDebugOptions');

                let traceId = "MatsSocketServer.DebugOptions_server.send_" + matsSocket.id(6);

                // These will become the server's initiation requested DebugOptions upon the subsequent 'send'
                matsSocket.debug = debugOptions;
                let initialMessage = "Message_" + matsSocket.id(20);

                matsSocket.endpoint("ClientSide.endpoint", (msg) => {
                    chai.assert.strictEqual(msg.data.number, Math.E);
                    chai.assert.strictEqual(msg.traceId, traceId);
                    extracted(msg, debugOptions);

                    return Promise.resolve({
                        string: msg.data.string + ":From_DebugOptions_test",
                        number: msg.data.number + Math.PI
                    })
                });

                // This terminator will get the final result from the Server
                matsSocket.terminator("ClientSide.terminator", (msg) => {
                    chai.assert.strictEqual(msg.data.number, Math.E + Math.PI);
                    chai.assert.strictEqual(msg.data.string, initialMessage + ":From_DebugOptions_test:RESOLVE");
                    chai.assert.strictEqual(msg.traceId, traceId);
                    done()
                });

                matsSocket.send("Test.server.request.direct", traceId, {
                    string: initialMessage,
                    number: Math.E
                })
            }
            it('Server initiated request should have debug object if user asks for it - with the all server-sent stuff filled.', function (done) {
                testServerInitiatedRequest_DebugOptions(mats.DebugOption.NODES | mats.DebugOption.TIMESTAMPS, done);
            });
            it('Server initiated request should have debug object if user asks for just nodes - with just the nodes filled.', function (done) {
                testServerInitiatedRequest_DebugOptions(mats.DebugOption.NODES, done);
            });
            it('Server initiated request should have debug object if user asks for just timestamps - with just the timestamps filled.', function (done) {
                testServerInitiatedRequest_DebugOptions(mats.DebugOption.TIMESTAMPS, done);
            });

            function serverInitiatedSend_NoDebugOptions(done, debugOptions) {
                // Set special userId that gives us all DebugOptions
                setAuth('enableAllDebugOptions');

                let traceId = "MatsSocketServer.DebugOptions_server.send_" + matsSocket.id(6);

                matsSocket.terminator("ClientSide.terminator", (msg) => {
                    chai.assert.strictEqual(msg.data.number, Math.E);
                    chai.assert.strictEqual(msg.traceId, traceId + ":SentFromThread");

                    // This is a server-initiated message, so these should be undefined in MessageEvent
                    chai.assert.isUndefined(msg.clientRequestTimestamp);
                    chai.assert.isUndefined(msg.roundTripMillis);

                    // :: Assert debug
                    // Since this is server-initiated, '0' will not give debug object - i.e. the server does not have difference between '0' and 'undefined'.
                    chai.assert.isUndefined(msg.debug);
                    done()
                });

                // :: These will become the server's initiation requested DebugOptions upon the subsequent 'send'
                // First set it to "all the things!"
                matsSocket.debug = 255;

                matsSocket.send("Test.single", "DebugOptions_set_to_all", {
                    number: Math.E
                });

                // Then set it to none (0 or undefined), which should result in NO debug object from server-initiated messages
                matsSocket.debug = debugOptions;

                matsSocket.send("Test.server.send.thread", traceId, {
                    number: Math.E
                })

            }
            it('Server initiated send should NOT have debug object if user ask for \'0\'.', function (done) {
                serverInitiatedSend_NoDebugOptions(done, 0);
            });
            it('Server initiated send should NOT have debug object if user ask for \'undefined\'.', function (done) {
                serverInitiatedSend_NoDebugOptions(done, undefined);
            });

        })
    });
}));

File: matssocket-client-dart/test/integration_push.dart
import 'dart:async';
import 'dart:math' as math;

import 'package:logging/logging.dart';
import 'package:matssocket/matssocket.dart';
import 'package:test/test.dart';

import 'lib/env.dart';

void main() {
  configureLogging();

  final _logger = Logger('integration_push');

  group('MatsSocket integration tests of Server-side send/request ("push")', () {
    late MatsSocket matsSocket;

    void setAuth(
        [String userId = 'standard',
        Duration duration = const Duration(seconds: 20),
        roomForLatencyMillis = const Duration(seconds: 10)]) {
      var now = DateTime.now();
      var expiry = now.add(duration);
      matsSocket.setCurrentAuthorization(
          'DummyAuth:$userId:${expiry.millisecondsSinceEpoch}', expiry, roomForLatencyMillis);
    }

    setUp(() => matsSocket = createMatsSocket());

    tearDown(() async  {
      await matsSocket.close('Test done');
      _logger.info('=========== Closed MatsSocket [${matsSocket.matsSocketInstanceId}] ===========');
    });

    group('MatsSocketServer.send()', () {
      // Set a valid authorization before each request
      setUp(setAuth);

      test('Send a message to Server, which responds by sending a message to terminator at Client (us!), directly in the MatsStage', () async {
        var traceId = 'MatsSocketServer.send_test_${id(6)}';
        var reply = matsSocket.terminator('ClientSide.terminator').first;
        await matsSocket.send('Test.server.send.matsStage', traceId, {'number': math.e});
        var msg = await reply;
        expect(msg.data['number'], equals(math.e));
        expect(msg.traceId, contains(':SentFromMatsStage'));
      });

      test('Send a message to Server, which responds by sending a message to terminator at Client (us!), in a separate Thread', () async {
        var traceId = 'MatsSocketServer.send_test_${id(6)}';
        var reply = matsSocket.terminator('ClientSide.terminator').first;

        await matsSocket.send('Test.server.send.thread', traceId, {'number': math.e});

        var msg = await reply;
        expect(msg.data['number'], equals(math.e));
        expect(msg.traceId, contains(':SentFromThread'));
      });
    });

    group('MatsSocketServer.request()', () {
      // Set a valid authorization before each request
      setUp(setAuth);

      Future doTest(String startEndpoint, bool resolveReject) async {
        var traceId = 'MatsSocketServer.send_test_${id(6)}';

        var initialMessage = 'Message_${id(20)}';

        // This endpoint will get a request from the Server, to which we respond - and the server will then send the reply back to the Terminator below.
        matsSocket.endpoint('ClientSide.endpoint', (messageEvent) {
          _logger.info('[sid:${matsSocket.sessionId}] We received a request to [ClientSide.endpoint], which we will [${resolveReject ? 'resolve' : 'reject'}], then wait for our terminator to be called');
          expect(messageEvent.traceId, equals(traceId));
          var promise = Completer();
          // Resolve it a tad later, to "emulate" some kind of processing
          Timer.run(() {
            var data = messageEvent.data;
            var msg = {'string': "${data['string']}:From_IntegrationEndpointA", 'number': data['number'] + math.pi};
            // ?: Resolve or Reject based on param
            if (resolveReject) {
              promise.complete(msg);
            } else {
              promise.completeError(msg);
            }
          });
          return promise.future;
        });

        // This terminator will get the final result from the Server
        var terminator = matsSocket.terminator('ClientSide.terminator');

        // Here we send the message that starts the cascade
        _logger.info('[sid:${matsSocket.sessionId}] Sending to [$startEndpoint], thus triggering the server to send a request');
        await matsSocket.send(startEndpoint, traceId, {'string': initialMessage, 'number': math.e});
        _logger.info('[sid:${matsSocket.sessionId}] Sending ACKed on server, the server should have sent a request, the client shall reply, and then the server should send to our terminator, which we are waiting for now');

        var msg = await terminator.first;
        _logger.info('[sid:${matsSocket.sessionId}] Terminator message received [$msg}], assrting content');
        expect(msg.data['number'], math.e + math.pi);
        expect(msg.data['string'], contains('From_IntegrationEndpointA'));
        expect(msg.data['string'], contains((resolveReject ? 'RESOLVE' : 'REJECT')));
        expect(msg.traceId, traceId);
      }

      test('Send a message to the Server, which responds by directly doing a Server-to-Client request (thus coming back here!), resolves, and when this returns to Server, sends it back to a Client Terminator', () async {
        await doTest('Test.server.request.direct', true);
      });

      test('Send a message to the Server, which responds by directly doing a Server-to-Client request (thus coming back here!), rejects, and when this returns to Server, sends it back to a Client Terminator', () async {
        await doTest('Test.server.request.direct', false);
      });

      test('Send a message to the Server, which responds by - in a Mats terminator - doing a Server-to-Client request (thus coming back here!), resolve, and when this returns to Server, sends it back to a Client Terminator', () async {
        await doTest('Test.server.request.viaMats', true);
      });

      test('Send a message to the Server, which responds by - in a Mats terminator - doing a Server-to-Client request (thus coming back here!), rejects, and when this returns to Server, sends it back to a Client Terminator', () async {
        await doTest('Test.server.request.viaMats', false);
      });
    });

    group('Server initiated sends and requests - with DebugOptions', () {
      void extracted(MessageEvent msg, int debugOptions) {
        // This is a server-initiated message, so these should be undefined in MessageEvent
        expect(msg.clientRequestTimestamp, isNull);
        expect(msg.roundTripMillis, isNull);

        // :: Assert debug
        // It should be present in MessageEvent, since it should be in the message from the server
        expect(msg.debug, isNotNull);
        // These should be set
        if ((debugOptions & DebugOption.TIMESTAMPS.flag) > 0) {
          expect(msg.debug!.serverMessageCreated, isA<DateTime>());
          expect(msg.debug!.messageSentToClient, isA<DateTime>());
        } else {
          expect(msg.debug!.serverMessageCreated, isNull);
          expect(msg.debug!.messageSentToClient, isNull);
        }
        if ((debugOptions & DebugOption.NODES.flag) > 0) {
          expect(msg.debug!.serverMessageCreatedNodename, isNotNull);
          expect(msg.debug!.messageSentToClientNodename, isNotNull);
        } else {
          expect(msg.debug!.serverMessageCreatedNodename, isNull);
          expect(msg.debug!.messageSentToClientNodename, isNull);
        }
        // While all other should not be set
        expect(msg.debug!.clientMessageSent, isNull);
        expect(msg.debug!.clientMessageReceived, isNull);
        expect(msg.debug!.clientMessageReceivedNodename, isNull);
        expect(msg.debug!.matsMessageSent, isNull);
        expect(msg.debug!.matsMessageReplyReceived, isNull);
        expect(msg.debug!.matsMessageReplyReceivedNodename, isNull);

        // This is not a request from Client, so there is no requested debug options.
        expect(msg.debug!.requestedDebugOptions, isNull);
        // This is the resolved from what we asked server to use above, and what we are allowed to.
        expect(msg.debug!.resolvedDebugOptions, equals(matsSocket.debug));
      }

      Future testServerInitiatedSend_DebugOptions(int debugOptions) async {
        // Set special userId that gives us all DebugOptions
        setAuth('enableAllDebugOptions');

        var traceId = 'MatsSocketServer.DebugOptions_server.send_${id(6)}';

        // These will become the server's initiation requested DebugOptions upon the subsequent 'send'
        matsSocket.debug = debugOptions;
        var terminator = matsSocket.terminator('ClientSide.terminator');
        await matsSocket.send('Test.server.send.thread', traceId, {'number': math.e});

        var msg = await terminator.first;
        expect(msg.data['number'], math.e);
        expect(msg.traceId, '$traceId:SentFromThread');

        extracted(msg, debugOptions);
      }

      test('Server initiated send should have debug object if user asks for it - with the all server-sent stuff filled.', () async {
        await testServerInitiatedSend_DebugOptions(DebugOption.NODES.flag | DebugOption.TIMESTAMPS.flag);
      });
      test('Server initiated send should have debug object if user asks for just nodes - with just the nodes filled.', () async {
        await testServerInitiatedSend_DebugOptions(DebugOption.NODES.flag);
      });
      test('Server initiated send should have debug object if user asks for just timestamps - with just the timestamps filled.', () async {
        await testServerInitiatedSend_DebugOptions(DebugOption.TIMESTAMPS.flag);
      });

      Future testServerInitiatedRequest_DebugOptions(int debugOptions) async {
        // Set special userId that gives us all DebugOptions
        setAuth('enableAllDebugOptions');

        var traceId = 'MatsSocketServer.DebugOptions_server.send_${id(6)}';

        // These will become the server's initiation requested DebugOptions upon the subsequent 'send'
        matsSocket.debug = debugOptions;
        var initialMessage = 'Message_${id(20)}';

        matsSocket.endpoint('ClientSide.endpoint', (msg) {
          expect(msg.data['number'], equals(math.e));
          expect(msg.traceId, equals(traceId));
          extracted(msg, debugOptions);

          return Future.value(
              {'string': "${msg.data['string']}:From_DebugOptions_test", 'number': msg.data['number'] + math.pi});
        });

        // This terminator will get the final result from the Server
        var terminator = matsSocket.terminator('ClientSide.terminator');

        await matsSocket.send('Test.server.request.direct', traceId, {'string': initialMessage, 'number': math.e});

        var msg = await terminator.first;
        expect(msg.data['number'], equals(math.e + math.pi));
        expect(msg.data['string'], equals('$initialMessage:From_DebugOptions_test:RESOLVE'));
        expect(msg.traceId, equals(traceId));
      }

      test('Server initiated request should have debug object if user asks for it - with the all server-sent stuff filled.', () async {
        await testServerInitiatedRequest_DebugOptions(DebugOption.NODES.flag | DebugOption.TIMESTAMPS.flag);
      });
      test('Server initiated request should have debug object if user asks for just nodes - with just the nodes filled.', () async {
        await testServerInitiatedRequest_DebugOptions(DebugOption.NODES.flag);
      });
      test('Server initiated request should have debug object if user asks for just timestamps - with just the timestamps filled.', () async {
        await testServerInitiatedRequest_DebugOptions(DebugOption.TIMESTAMPS.flag);
      });

      Future serverInitiatedSend_NoDebugOptions(int? debugOptions) async {
        // Set special userId that gives us all DebugOptions
        setAuth('enableAllDebugOptions');

        var traceId = 'MatsSocketServer.DebugOptions_server.send_${id(6)}';
        var terminator = matsSocket.terminator('ClientSide.terminator');

        // :: These will become the server's initiation requested DebugOptions upon the subsequent 'send'
        // First set it to "all the things!"
        matsSocket.debug = 255;

        await matsSocket.send('Test.single', 'DebugOptions_set_to_all', {'number': math.e});

        // Then set it to none (0 or undefined), which should result in NO debug object from server-initiated messages
        matsSocket.debug = debugOptions;

        await matsSocket.send('Test.server.send.thread', traceId, {'number': math.e});

        var msg = await terminator.first;

        expect(msg.data['number'], equals(math.e));
        expect(msg.traceId, equals(traceId + ':SentFromThread'));

        // This is a server-initiated message, so these should be undefined in MessageEvent
        expect(msg.clientRequestTimestamp, isNull);
        expect(msg.roundTripMillis, isNull);

        // :: Assert debug
        // Since this is server-initiated, '0' will not give debug object - i.e. the server does not have difference between '0' and 'undefined'.
        expect(msg.debug, isNull);
      }

      test('Server initiated send should NOT have debug object if user ask for 0.', () async {
        await serverInitiatedSend_NoDebugOptions(0);
      });
      test('Server initiated send should NOT have debug object if user ask for \'undefined\'.', () async {
        await serverInitiatedSend_NoDebugOptions(null);
      });
    });
  });
}
